src/ATen/native/xpu/sycl/BatchNormKernels.cpp

```diff
@@ -1263,6 +1263,147 @@ struct BatchNormTransformInputKernelFunctor {
   stat_accscalar_t epsilon_;
 };
 
+template <
+    int VEC_SIZE,
+    typename input_scalar_t,
+    typename stat_scalar_t,
+    typename stat_accscalar_t,
+    bool train,
+    typename index_t>
+struct BatchNormTransformInputVectorizedKernelFunctor {
+  void operator()(sycl::nd_item<2> item) const {
+    index_t plane = item.get_group(1);
+
+    if (plane >= input_.size(1)) {
+      return;
+    }
+
+    stat_accscalar_t gamma = weight_.size(0) > 0
+        ? static_cast<stat_accscalar_t>(weight_[plane])
+        : static_cast<stat_accscalar_t>(1);
+    stat_accscalar_t beta = bias_.size(0) > 0
+        ? static_cast<stat_accscalar_t>(bias_[plane])
+        : static_cast<stat_accscalar_t>(0);
+    stat_accscalar_t mean = static_cast<stat_accscalar_t>(mean_[plane]);
+    stat_accscalar_t invstd;
+    if constexpr (train) {
+      invstd = var_or_invstd_[plane];
+    } else {
+      invstd =
+          static_cast<stat_accscalar_t>(1) /
+          device_sqrt(
+              static_cast<stat_accscalar_t>(var_or_invstd_[plane]) + epsilon_);
+    }
+
+    index_t bs = input_.size(0);
+    index_t fs = input_.size(2);
+
+    index_t bstep = item.get_local_range(0) * item.get_group_range(0);
+    for (index_t batch = item.get_global_id(0); batch < bs; batch += bstep) {
+      auto o = output_[batch][plane];
+      auto i = input_[batch][plane];
+
+      for (index_t feature_vec_begin = item.get_local_id(1) * VEC_SIZE;
+           feature_vec_begin < fs;
+           feature_vec_begin += VEC_SIZE * item.get_local_range(1)) {
+        auto remaining = fs - feature_vec_begin;
+        if (remaining < VEC_SIZE) {
+          for (index_t idx = 0; idx < remaining; ++idx) {
+            index_t feature = feature_vec_begin + idx;
+            o[feature] = static_cast<input_scalar_t>(
+                gamma * (i[feature] - mean) * invstd + beta);
+          }
+        } else {
+          using vec_t = memory::aligned_vector<input_scalar_t, VEC_SIZE>;
+          vec_t vec;
+#pragma unroll
+          for (int vt = 0; vt < VEC_SIZE; ++vt) {
+            index_t feature = feature_vec_begin + vt;
+            vec[vt] = static_cast<input_scalar_t>(
+                gamma * (i[feature] - mean) * invstd + beta);
```

Comment:
Why not using vectorization for inputs? I see `vt` is contiguous across the iteration.

Reply:
If we know the `vt` size, we can calculate offset with linear id 0, 3, 7, 11? if the `vt` size is 4?

Reply:
Apply a coalescing manner for the input access would be quite complex, let we discuss it later.